### PR TITLE
self validating migration

### PR DESCRIPTION
**Pull request by jsign**

### CODE REVIEW

File: core/overlay/conversion.go [hunk 1]
```diff
@@ -319,7 +370,7 @@ func OverlayVerkleTransition(statedb *state.StateDB, root common.Hash, maxMovedC
 				// processing the storage for that account where we left off.
 				// If the entire storage was processed, then the iterator was
 				// created in vain, but it's ok as this will not happen often.
```
Comment (jsign) [inline core/overlay/conversion.go:372]: Changes from here until the end of the file are just to always migrate "full accounts" -- just to discard any half-migrated situations.

File: core/state_processor.go [hunk 2]
```diff
@@ -104,7 +104,7 @@ func (p *StateProcessor) Process(block *types.Block, statedb *state.StateDB, cfg
 	}
 	parent := p.bc.GetHeaderByHash(header.ParentHash)
 	if err := overlay.OverlayVerkleTransition(state, parent.Root, p.config.OverlayStride); err != nil {
-		log.Error("error performing the transition", "err", err)
+		return nil, nil, 0, fmt.Errorf("error performing transition: %s", err)
```
Comment (jsign) [inline core/state_processor.go:107]: This was bad!

File: core/overlay/conversion.go [hunk 3]
```diff
@@ -86,13 +92,58 @@ func newBranchKey(addr []byte, treeIndex *uint256.Int) branchKey {
 	return sk
 }
 
+type migratedSS struct {
+	slotNumber []byte
+	slotValue  []byte
+}
+
 func (kvm *keyValueMigrator) addStorageSlot(addr []byte, slotNumber []byte, slotValue []byte) {
+	kvm.migratedStorageSlots[string(addr)] = migratedSS{slotNumber: slotNumber, slotValue: slotValue}
```
Comment (jsign) [inline core/overlay/conversion.go:101]: Save this for further verification.

File: core/overlay/conversion.go [hunk 4]
```diff
@@ -86,13 +92,58 @@ func newBranchKey(addr []byte, treeIndex *uint256.Int) branchKey {
...
 	treeIndex, subIndex := utils.GetTreeKeyStorageSlotTreeIndexes(slotNumber)
 	leafNodeData := kvm.getOrInitLeafNodeData(newBranchKey(addr, treeIndex))
 	leafNodeData.Values[subIndex] = slotValue
 }
 
+func (kvm *keyValueMigrator) validate(tree *trie.TransitionTrie) error {
+	log.Info("Validating accounts", "count", len(kvm.migratedAccounts))
+	// 06012c8cf97bead5deae237070f9587f8e7a266d
```
Comment (jsign) [inline core/overlay/conversion.go:109]: This is the account that I see has an inconsistent balance in the migration.

File: core/overlay/conversion.go [hunk 5]
```diff
@@ -86,13 +92,58 @@ func newBranchKey(addr []byte, treeIndex *uint256.Int) branchKey {
...
+		if !bytes.Equal(got.CodeHash, acc.CodeHash) {
+			return fmt.Errorf("code hash doesn't match")
+		}
+		if got.Nonce != acc.Nonce {
+			return fmt.Errorf("nonce doesn't match")
+		}
+		// if got.Root != acc.Root {
+		// 	return fmt.Errorf("root doesn't match")
+		// }
+	}
```
Comment (jsign) [inline core/overlay/conversion.go:128]: Check all the migrated accounts.

File: core/overlay/conversion.go [hunk 6]
```diff
@@ -86,13 +92,58 @@ func newBranchKey(addr []byte, treeIndex *uint256.Int) branchKey {
...
+	for addr, ss := range kvm.migratedStorageSlots {
+		log.Info("Checking storage slot", "addr", hex.EncodeToString([]byte(addr)), "storage slot", hex.EncodeToString(ss.slotNumber))
+		got, err := tree.GetStorage(common.BytesToAddress([]byte(addr)), ss.slotNumber)
+		if err != nil {
+			return fmt.Errorf("failed to get storage slot: %s", err)
+		}
+		if !bytes.Equal(got, ss.slotValue) {
+			return fmt.Errorf("storage slot value doesn't match")
+		}
+	}
```
Comment (jsign) [inline core/overlay/conversion.go:140]: Check all the migrated storage slots.

File: core/state_processor.go (hunk 2 above)
Comment (gballet) [reply core/state_processor.go:107]: ah yes, definitely a merge typo

**Comment by jsign** [comment]
It was helpful to discover some bugs, so now closing.

**Pull request closed without merging**